### DISCUSSION TITLE
Persistence

### DISCUSSION THREAD

**Discussion by dniminenn** [Kilvoctu/aiyabot#123] [General]
Especially for the clipboard button, what would we need in order to make the button persistent?

**Comment by Kilvoctu** [comment]
Heya, it's 2nd incomplete entry on #85, right under gui (which I'm working on currently).  

Hypothetically, store interaction object of every message into sql database alongside its respective message id. Have buttons able to know what message id they're attached to, reference it in db and grab interaction object. This object should have what the button needs in order to function, I believe.  
That said, I'm not sure this approach actually would work, but it seems plausible. Is probably more complicated than that.  

This issue is one of the few times I ever requested help (in PyCord support Discord), and I was bluntly told it's not possible. Not sure I agree with the answer, but I just haven't had time to try.

**Comment by solstice-r** [comment]
Don't Discord interactions expire forever after 15 minutes? If there's no way to bypass that limit, that's probably why you were told it's impossible to make the buttons persistent.

**Reply by Kilvoctu** [reply]
My memory is a bit fuzzy, since it's been a while since I looked into the specifics behind it; the buttons were added ~4 months ago.

The buttons are placed into a view, and the view is what's attached to each message. I think I make the view persistent by initializing it separately when setting up the /draw command options: https://github.com/Kilvoctu/aiyabot/blob/4338c1e24e5a523416fc99304b041726d2c565e9/core/stablecog.py#L29-L31  
Since otherwise, I'm not sure why past me added that, since buttons still work without it. Then on each view in viewhandler.py, the timeout is set to none.  

It seems to work, since I leave my personal AIYA up for days at a time and buttons stay functional. By default, I think there is the timer as you mentioned, since otherwise probably a bot would be bogged down over time by remembering so many interactions. And for most use cases, there's no reason to keep buttons around that long.

---

I think on PyCord Discord, my question was something like "How to keep buttons/view working after a restart". Checking there again now, I see someone else asked the same question and they were linked to [this example](https://github.com/Pycord-Development/pycord/blob/master/examples/views/persistent.py) with no explanation, and it doesn't solve the issue lol. So I feel like the support there is just sorta bad, in my opinion.  
At least, I do see in the example a comment line mentioning a db, but no example on how to do it.

**Comment by dniminenn** [comment]
I would consider adding a command to invoke pnginfo. The pnginfo command could take a url or upload as parameter and generate a /draw command just like the clipboard button, with the added benefit of not requiring persistence...

Edit: this isn't a feature request lol, just a potential nice to have within the context of this discussion

**Reply by Kilvoctu** [reply]
I could add a pnginfo fallback, sure. I'm very resistant on adding slash commands, so I'll likely integrate it into /identify. I'll look into it now.

That said, I'd much prefer that buttons work reliably. Buttons that may or may not work seemingly randomly isn't a great end user experience.

**Reply by Kilvoctu** [reply]
> Edit: this isn't a feature request lol, just a potential nice to have within the context of this discussion

Me being receptive to feedback is why AIYA is developed past week 1, lol. If it's something I think is cool or useful, I'll want to try to find a way to add it.

**Reply by Kilvoctu** [reply]
I added a very simple png info implementation with #129.  
It only shows the PNG info like it does in Web UI, but people will have *something* they can do until buttons can be persistent.

It's a bit tricky to parse that info into a command for copying. Probably would have to drop some parameters (like model) and make a note that users need to enter those options manually.

**Reply by dniminenn** [reply]
> That said, I'd much prefer that buttons work reliably. Buttons that may or may not work seemingly randomly isn't a great end user experience.

Agreed, I like to add that you are right, /identify would be the ideal place to incorporate pnginfo functionality. It already gives you two options, Tags and Normal, it would make sense to use the same init_image parameter and add a pnginfo option. Free time coming up in a few weeks

**Reply by Kilvoctu** [reply]
Oh you're right adding pnginfo to "**phrasing**" makes more sense 🤔

edit: did it

**Reply by dniminenn** [reply]
Sorry, this particular discussion grew quite a bit in scope haha!!

**Reply by Kilvoctu** [reply]
> edit: did it

Oops, had it commited to the wrong branch. Now actually did it for real #130

**Comment by danstis** [comment]
I did a little reading on this issue and my take away was that the use of persistent views is the correct method to have the buttons working across restarts. 
From what I see, you already have custom ids on all the buttons, and you have an expiry set to none. 
Potentially you might need to have the button_x custom id modified because it is used on 2 separate buttons. 

After this, it seems that the only thing required is to call the view with no arguments in the on_ready function like in the example: https://github.com/Pycord-Development/pycord/blob/71db6f6de5df8b7bb6c785ecabd5b5dc28089cdf/examples/views/persistent.py#L56

**Reply by Kilvoctu** [reply]
I do already have the view in the on_ready function, but it needs its respective argument in order to work.  

Just now, I messed around with the code for a bit, and I think I remember now the persistent view ensures that on restarts the buttons themselves can still work. I removed my on_ready function, and after a restart the buttons simply don't respond at all (instead of AIYA saying they broke).  
The persistent view helps for a simple code like this one I tested. This works across restarts: 
```
@discord.ui.button(custom_id="button_hello", emoji="👍")
async def button_roll(self, button, interaction):
    await interaction.response.send_message("Hello.", ephemeral=True)
```

I guess it's something else actually that I need to store into a db for the buttons to retrieve. In the console log, this is the error from trying to use a button after restart:
`The clipboard button broke: 'StableCog' object is not subscriptable`

**Reply by danstis** [reply]
Ahh, yeah got ya. 
Quite a pain.

**Reply by dniminenn** [reply]
Yeah, I've researched this a bit and I'm quite positive the only way to persist the clipboard button is to store the button state in a database. I suggest sqlite3 and keeping a sidecar db file in the resources directory of the bot. I don't mind working on this in the coming weeks if you need to delegate. sqlite has minimal dependencies and the bot owner needn't worry about configuring a database

**Reply by Kilvoctu** [reply]
Coincidentally, I was looking into it a bit and came to the same conclusions lol. I have a basic db setup done, went ahead and pushed it to a different branch to look at: https://github.com/Kilvoctu/aiyabot/commit/9798b31119ac7627860d40845e507f08ab47d1b3 and it seemed to at least make an entry 
![image](https://user-images.githubusercontent.com/2993060/221388049-270bbf3b-26a9-481f-883c-833dc343154f.png)

Think I'll be able to handle this, but I appreciate the offer 👍👍.  
The GUI project can be put on hold for now, as it's mostly a convenience not a priority.

**Reply by Kilvoctu** [reply]
I've got some headway into making buttons persistent, but have encountered two issues:  

Can't find a way to store ctx (discord.ApplicationContext) into a db. This is used in the queue_object to know who to reply to and what channel to send the output to. I think I have some workarounds for this, but the next issue is more pressing.  

The buttons (View, actually) do not seem to know what message they're attached to after a restart. So I can't do `self.message.id` or `self.message.content` (to read what's in the message).  
I'm not sure what exactly a Discord View can remember after a restart. I was playing around with uuid, so if each View could somehow receive and remember this through restarts, then it forgetting the message is inconsequential.  

edit: I'll let this issue marinate in my head for a bit and take a look at Discord's context menu feature to see what that's all about.
